activejob/lib/active_job/exceptions.rb

```diff
@@ -44,12 +44,22 @@ module ClassMethods
       #  end
       def retry_on(*exceptions, wait: 3.seconds, attempts: 5, queue: nil, priority: nil)
         rescue_from(*exceptions) do |error|
+          payload = {
+            job: self,
+            adapter: self.class.queue_adapter,
+            error: error,
+          }
+
           if executions < attempts
-            logger.error "Retrying #{self.class} in #{wait} seconds, due to a #{error.class}. The original exception was #{error.cause.inspect}."
-            retry_job wait: determine_delay(wait), queue: queue, priority: priority
+            ActiveSupport::Notifications.instrument("enqueue_retry.active_job", payload) do
+              logger.error "Retrying #{self.class} in #{wait} seconds, due to a #{error.class}. The original exception was #{error.cause.inspect}."
+              retry_job wait: determine_delay(wait), queue: queue, priority: priority
+            end
           else
             if block_given?
-              yield self, error
+              ActiveSupport::Notifications.instrument("retry_stopped.active_job", payload) do
+                yield self, error
+              end
             else
               logger.error "Stopped retrying #{self.class} due to a #{error.class}, which reoccurred on #{executions} attempts. The original exception was #{error.cause.inspect}."
               raise error
```

Comment:
I think it is still worth to notify in this case. For this case we can just use the non-block format so we notify before the exception is raised.